src/Model.integration.test.ts

```diff
@@ -0,0 +1,244 @@
+  });
+
+  it<TestContext>('rejects the data in optimistic() and rolls back the changes if back-end rejects the data', async ({
+    ably,
+    model,
+    channelName,
+    eventData,
+  }) => {
+    await model.sync(eventData.data);
+    let subscription = new Subject<void>();
+    const subscriptionCalls = getEventPromises(subscription, 3);
+    const subscriptionSpy = vi.fn(() => subscription.next());
+    model.subscribe(subscriptionSpy);
+
+    await model.optimistic(eventData);
+    expect(model.data.optimistic).toEqual(eventData.data);
+    const channel = ably.channels.get(channelName);
+    await channel.publish({
+      data: eventData.data,
+      name: 'update',
+      extras: {
+        headers: {
+          'x-ably-models-reject': true,
+        },
+      },
+    });
+
+    await subscriptionCalls[1];
+
+    expect(subscriptionSpy).toHaveBeenCalledTimes(2);
+    expect(subscriptionSpy).toHaveBeenNthCalledWith(2, null, eventData.data);
+    expect(model.data.confirmed).toEqual(undefined);
+  });
+
+  it<TestContext>('rejects the data and rolls back the changes if optimistic() timeouts', async ({
+    model,
+    eventData,
+  }) => {
+    await model.sync(eventData.data);
+    const [confirmation] = await model.optimistic(eventData, { timeout: 10 });
+    expect(model.data.optimistic).toEqual(eventData.data);
+
+    try {
+      await confirmation;
+    } catch (e) {
+      expect(e).toEqual(new Error('timed out waiting for event confirmation'));
+      expect(model.data.optimistic).toEqual(undefined);
+      expect(model.data.confirmed).toEqual(undefined);
+    }
+  });
+
+  it<TestContext>('rebases optimistic data on top of subsequent event data', async ({
+    ably,
+    model,
+    channelName,
+    eventData,
+  }) => {
+    const channel = ably.channels.get(channelName);
+    const blazinglyFastEvent = {
+      data: {
+        comment: "I'm blazingly fast!",
+      },
+      mutationId: 'some-id-2',
+      name: 'updateComment',
+    };
+    await model.sync(eventData.data);
+    let subscription = new Subject<void>();
+    const subscriptionCalls = getEventPromises(subscription, 5);
+    const subscriptionSpy = vi.fn(() => subscription.next());
+    model.subscribe(subscriptionSpy);
+
+    const [confirmation] = await model.optimistic(eventData);
+    expect(model.data.optimistic).toEqual(eventData.data);
+
+    await channel.publish({
+      ...blazinglyFastEvent,
+      extras: {
+        headers: {
+          'x-ably-models-event-uuid': blazinglyFastEvent.mutationId,
+        },
+      },
+    });
+
+    await subscriptionCalls[2];
+    expect(model.data.confirmed).toEqual({
+      ...blazinglyFastEvent.data,
+    });
+
+    await channel.publish({
+      ...eventData,
+      extras: {
+        headers: {
+          'x-ably-models-event-uuid': eventData.mutationId,
+        },
+      },
+    });
+
+    await subscriptionCalls[5];
+    await confirmation;
+
+    expect(model.data.confirmed).toEqual({
```

Comment:
I'm not sure this is doing what we expect:

1. The initial confirmed data is `eventData.data` as that is what is passed to `sync`
2. We then apply the same event optimistically (no detectable change in the data that we can assert on)
3. We then publish the confirmed update `blazinglyFastEvent`; `eventData.data` will be rebased
4. We then confirm `eventData.data`

We should fix 2. (this issue is present in other tests per prev comment)

We should also avoid using the spread operator here to merge in object keys, as this doesn't allow us to tell the order in which updates were applied (since we are using distinct keys so both will be present in the resulting object regardless of the order the updates are applied in). Instead it would be better to use e.g. an array, which allows us to see the order, i.e. that `blazinglyFastEvent` comes first and `eventData` comes second.

Reply:
I was lazy and didn't change the sync data into an array but I did change the object properties, so it's more visible as to what data came in first and last. Let me know if you don't like it, I'll change that into an array. 🙃